modules/core/src/main/java/org/apache/fluo/core/impl/TransactionImpl.java

```diff
@@ -1141,178 +1210,340 @@ private void beginSecondCommitPhase(CommitData cd, Stamp commitStamp) throws Exc
+        // TODO async
+        TxInfo txInfo = TxInfo.getTransactionInfo(env, cd.prow, cd.pcol, startTs);
 
-            // TODO async
-            TxInfo txInfo = TxInfo.getTransactionInfo(env, cd.prow, cd.pcol, startTs);
-
-            switch (txInfo.status) {
-              case COMMITTED:
-                if (txInfo.commitTs != commitTs) {
-                  throw new IllegalStateException(
-                      cd.prow + " " + cd.pcol + " " + txInfo.commitTs + "!=" + commitTs);
-                }
-                ms = Status.ACCEPTED;
-                break;
-              case LOCKED:
-                // TODO async
-                ms = cd.cw.write(delLockMutation).getStatus();
-                break;
-              default:
-                ms = Status.REJECTED;
+        switch (txInfo.status) {
+          case COMMITTED:
+            if (txInfo.commitTs != commitTs) {
+              throw new IllegalStateException(
+                  cd.prow + " " + cd.pcol + " " + txInfo.commitTs + "!=" + commitTs);
             }
-          }
-
-          postCommitPrimary(cd, commitTs, ms);
+            ms = Status.ACCEPTED;
+            break;
+          case LOCKED:
+            // TODO async
+            ConditionalMutation delLockMutation = result.getMutation();
+            ms = cd.cw.write(delLockMutation).getStatus();
+            break;
+          default:
+            ms = Status.REJECTED;
         }
-      };
+      }
+      Result newResult = new Result(ms, result.getMutation(), result.getTabletServer());
+      return Collections.singletonList(newResult).iterator();
+    }
 
-      env.getSharedResources().getSyncCommitExecutor().execute(task);
-    } else {
-      postCommitPrimary(cd, commitTs, mutationStatus);
+    @Override
+    public boolean processResults(CommitData cd, Iterator<Result> results) throws Exception {
+      Result result = Iterators.getOnlyElement(results);
+      return result.getStatus() == Status.ACCEPTED;
     }
+
+    @Override
+    CompletableFuture<Void> getFailureOp(CommitData cd) {
+      return CompletableFuture.runAsync(() -> {
+        cd.commitObserver.commitFailed(cd.getShortCollisionMessage());
+      }, env.getSharedResources().getSyncCommitExecutor());
+    }
+
   }
 
-  private void postCommitPrimary(CommitData cd, long commitTs, Status mutationStatus)
-      throws Exception {
-    if (mutationStatus != Status.ACCEPTED) {
-      cd.commitObserver.commitFailed(cd.getShortCollisionMessage());
-    } else {
-      if (stopAfterPrimaryCommit) {
-        cd.commitObserver.committed();
-      } else {
-        deleteLocks(cd, commitTs);
+  @VisibleForTesting
+  public boolean finishCommit(CommitData cd, Stamp commitStamp)
+      throws TableNotFoundException, MutationsRejectedException {
+    getStats().setCommitTs(commitStamp.getTxTimestamp());
+
+    CommitStep firstStep = new DeleteLocksStep();
+    firstStep.andThen(new FinishCommitStep());
+    firstStep.compose(cd);
```

Comment:
It would be good to add exception handling here.  Does not seem that exceptions are expected here.   Could add an exception handler that just prints or logs and unexpected exceptions.

```java
    CommitStep firstStep = new DeleteLocksStep();
    firstStep.andThen(new FinishCommitStep());
    firstStep.compose(cd).exceptionally(t -> {
      System.err.println("Unexpected exception in finish commit test method : ");
      t.printStackTrace();
      return null;
    });
```